geode-core/src/main/java/org/apache/geode/internal/cache/execute/FunctionContextImpl.java

```diff
@@ -37,20 +38,25 @@
 
   private String functionId = null;
 
+  private Cache cache = null;
+
   private ResultSender resultSender = null;
 
   private final boolean isPossDup;
 
   public FunctionContextImpl(final String functionId, final Object args,
       ResultSender resultSender) {
-    this.functionId = functionId;
-    this.args = args;
-    this.resultSender = resultSender;
-    this.isPossDup = false;
+    this(null, functionId, args, resultSender, false);
+  }
+
+  public FunctionContextImpl(final Cache cache, final String functionId, final Object args,
+      ResultSender resultSender) {
+    this(cache, functionId, args, resultSender, false);
   }
 
-  public FunctionContextImpl(final String functionId, final Object args, ResultSender resultSender,
-      boolean isPossibleDuplicate) {
+  public FunctionContextImpl(final Cache cache, final String functionId, final Object args,
+      ResultSender resultSender, boolean isPossibleDuplicate) {
```

Comment:
we can't remove GemFireCacheImpl.getInstance() for now, that's the way we are getting the instance to create the FunctionContext in the first place, why not just use it to avoid making more constructor changes. This just my 2 cents. I can go either way on this though.

The problem this changeset is trying to solve is that if external developer can not use GemfireCacheImpl.getInstance() to get the cache since it's an internal api, they can only use CacheFactory.getAnyIntance() and that's a synchronized static, so it requires a lock on CachFactory which creates some deadlocks.